### DISCUSSION TITLE
Make it easy to depend on artifacts that are binary compatible with transitive dependencies

### DISCUSSION THREAD

**Discussion by jgogstad** [sbt/sbt#6682] [Ideas]
Often when bringing in new dependencies, specifically when you bring in another module of a project that is already on the classpath by a transitive dependency, you want to just "follow the version of the artifact that is brought in transitively".

That is a clumsy statement, an example will make it clear

1. You depend on `"org.http4s" %% "http4s-circe"`, it brings in `"io.circe" %% "circe-core"` transitively
1. Circe is a multi-module module project and publishes another artifact that you want to depend on (say) `"io.circe" %% "circe-generic"`. You would like to "use whatever version is already on the classpath" for this dependency.

Why?

- Circe (in this example, but it goes for anything, especially Java libraries) doesn't always maintain binary compatibility. We're not interested in having the newest version of the artifact as much as having a compatible version of the artifact.
- Adding new dependencies is hard, you need to inspect the effective classpath in order to be sure to not break anything
- Upgrading dependencies is even worse. 
- Automated dependency upgrades, for instance with Scala Steward, becomes more involved. We don't want to bump versions that are tied to transitive dependencies

I know there's work being done to help with binary compatibility in [sbt-version-policy](https://github.com/scalacenter/sbt-version-policy), and that you do get eviction warnings whenever there's evictions.

To the eviction warnings, I will say that _for me_ at least, they suffer from broken window syndrome. I haven't found any practical way to keep the dependency graph free of eviction warnings—_especially_ for modules that depend on beasts such as Spark or Beam—the eviction warnings become "warnings you ignore because they're always there".

Very interested to get input on how to maintain a binary compatible classpath, and whether the proposed feature would be considered useful.

**Comment by eed3si9n** [comment]
> To the eviction warnings, I will say that for me at least, they suffer from broken window syndrome. 

I've removed eviction warnings in sbt 1.5.0 - https://eed3si9n.com/sbt-1.5.0. Instead we have eviction error. It shouldn't say anything unless it's sure, and when it's sure it will now block the `compile`.

If you want to add direct dependencies based on what you have in the graph, why not use sbt-dependency-graph? That's in-sourced into sbt since 1.4.0 - https://eed3si9n.com/sbt-1.4.0

**Reply by jgogstad** [reply]
Thanks for the reply.

> I've removed eviction warnings in sbt 1.5.0 - https://eed3si9n.com/sbt-1.5.0. Instead we have eviction error. It shouldn't say anything unless it's sure, and when it's sure it will now block the compile.

Ah, I didn't know that warnings were actually removed. Thanks for pointing it out.

> If you want to add direct dependencies based on what you have in the graph, why not use sbt-dependency-graph? That's in-sourced into sbt since 1.4.0 - https://eed3si9n.com/sbt-1.4.0

I do this manually whenever I want to add a direct dependency that "follows the version in graph", so automating that process with something like `resolvedVersion <organization> <module>` (or something) would solve my problem. Implementing that task doesn't sound like a huge endeavour, I'll give it a shot.

**Reply by eed3si9n** [reply]
I'd encourage you to iterate using plugin and see how people find it.
